cpp/src/join/hash_join.cuh

```diff
@@ -0,0 +1,336 @@
+    if((h_size_estimate > 0)
+       || (sample_probe_num_rows >= probe_table_num_rows))
+    {
+      break;
+    }
+
+    // If the size estimate is zero, then double the number of sampled rows in the probe
+    // table. Reduce the ratio of the number of probe rows sampled to the number of rows
+    // in the build table by the same factor
+    if(0 == h_size_estimate) {
+      constexpr size_type GROW_RATIO{2};
+      sample_probe_num_rows *= GROW_RATIO;
+      probe_to_build_ratio = static_cast<size_type>(std::ceil(static_cast<float>(probe_to_build_ratio)/GROW_RATIO));
+    }
+
+  } while(true);
+
+  e.set_value(h_size_estimate);
+  return std::make_unique<
+    experimental::scalar_type_t<size_type>>(
+        std::move(e));
+}
+
+template <JoinType join_type, typename index_type>
+std::enable_if_t<(join_type != JoinType::FULL_JOIN),
+std::unique_ptr<experimental::table>>
+get_base_hash_join_indices(
+    table_view const& left,
+    table_view const& right,
+    bool flip_join_indices,
+    cudaStream_t stream,
+    rmm::mr::device_memory_resource* mr) {
+
+  if ((join_type == JoinType::INNER_JOIN) && (right.num_rows() > left.num_rows())) {
+    return get_base_hash_join_indices<join_type, index_type>(right, left,  true, stream, mr);
+  }
+  if ((join_type == JoinType::LEFT_JOIN) && (right.num_rows() == 0)) {
+    rmm::device_buffer sequence_indices{
+      sizeof(index_type)*left.num_rows(), stream, mr};
+    thrust::device_ptr<index_type> sequence_indices_ptr(
+        static_cast<index_type*>(sequence_indices.data()));
+    thrust::sequence(
+        rmm::exec_policy(stream)->on(stream),
+        sequence_indices_ptr,
+        sequence_indices_ptr + left.num_rows(),
+        0);
+    rmm::device_buffer invalid_indices{
+      sizeof(index_type)*left.num_rows(), stream, mr};
+    thrust::device_ptr<index_type> inv_index_ptr(
+        static_cast<index_type*>(invalid_indices.data()));
+    thrust::fill(
+        rmm::exec_policy(stream)->on(stream),
+        inv_index_ptr,
+        inv_index_ptr + left.num_rows(),
+        JoinNoneValue);
+    return get_indices_table<index_type>(
+        std::move(sequence_indices), std::move(invalid_indices),
+        left.num_rows(), stream, mr);
+  }
+
+  //TODO : Add trivial left join case and exit early
+
+  using multimap_type = multimap_T<index_type>;
+
+  //TODO : attach stream to hash map class according to PR discussion #3272
+
+  auto build_table = table_device_view::create(right, stream);
+  const size_type build_table_num_rows{build_table->num_rows()};
+
+  // Probe with the left table
+  auto probe_table = table_device_view::create(left, stream);
+
+  // Hash table size must be at least 1 in order to have a valid allocation.
+  // Even if the hash table will be empty, it still must be allocated for the
+  // probing phase in the event of an outer join
+  size_t const hash_table_size =
+      std::max(compute_hash_table_size(build_table_num_rows), size_t{1});
```

Comment:
Is this necessary?  The only way compute_hash_table_size would be 0 would be if build_table_num_rows is 0.  I think all of the wrappers that call this underlying function abort early if the table is of size 0.

Reply:
Removed